### PR TITLE
Create region security policies rules

### PR DESCRIPTION
**Pull request by felipegc**
Fixes:
https://github.com/hashicorp/terraform-provider-google/issues/15319

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/develop/run-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_compute_region_security_policy_rule`
```

```release-note:enhancement
compute: added support for `user_defined_fields` to `google_compute_region_security_policy`
```

### CODE REVIEW

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+<% unless version == 'ga' -%>
+
+import (
+	"testing"
+	"github.com/hashicorp/terraform-provider-google/google/acctest"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccComputeRegionSecurityPolicyRule_regionSecurityPolicyRuleBasicUpdateExample(t *testing.T) {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:12]:
```suggestion
func TestAccComputeRegionSecurityPolicyRule_regionSecurityPolicyRuleBasicUpdate(t *testing.T) {
```

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    versioned_expr = "SRC_IPS_V1"
+    config {
+      src_ip_ranges = ["172.16.0.0/12"]
+    }
+  }
+}
+`, context)
+}
+
+func TestAccComputeRegionSecurityPolicyRule_regionSecurityPolicyRuleNetworkMatchUpdateExample(t *testing.T) {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:96]:
```suggestion
func TestAccComputeRegionSecurityPolicyRule_regionSecurityPolicyRuleNetworkMatchUpdate(t *testing.T) {
```

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_test.go.erb [hunk 3]
```diff
@@ -68,4 +68,103 @@ resource "google_compute_region_security_policy" "regionSecPolicy" {
 }
 `, context)
 }
+
+func TestAccComputeRegionSecurityPolicy_regionSecurityPolicyUserDefinedFieldsUpdateExample(t *testing.T) {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_compute_region_security_policy_test.go.erb:72]:
```suggestion
func TestAccComputeRegionSecurityPolicy_regionSecurityPolicyUserDefinedFieldsUpdate(t *testing.T) {
```

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 1 above)
Comment (rileykarson) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:12]: We should omit the `Example` suffix from handwritten tests- they're not testing docs examples.

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+  
+resource "google_compute_network_edge_security_service" "edge_sec_service" {
+  region          = "asia-southeast1"
+  name            = "tf-test-edgesec%{random_suffix}"
+  security_policy = google_compute_region_security_policy.policyddosprotection.self_link
+}
+  
+resource "google_compute_region_security_policy" "policynetworkmatch" {
+  region      = "asia-southeast1"
+  project     = "ciandt-cloudarmor-gcam"
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:162]: Is this supposed to be a different project than the rest of the configuration? (same with others below)

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 1 above)
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb]: Done

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_test.go.erb (hunk 3 above)
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_test.go.erb]: Done

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 4 above)
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:162]: No. Thanks for catching this.
Done.

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 2 above)
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb]: Done

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 4 above)
Comment (rileykarson) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:162]: These didn't get updated in your last commit, just the test names!

Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb]: Done

**Comment by rileykarson** [comment]
Adding new reviewer as I'll be OOO

**Comment by felipegc** [comment]
@rileykarson if you still not OOO, could you please run the test again? The resource "google_compute_network_edge_security_service" can exist only one per region so I moved one test to another region to test if the remaining test passes. Thanks

**Comment by zli82016** [comment]
The tests is running.

**Comment by felipegc** [comment]
Hi @zli82016, I hope you are doing well.
I see that the test is still failing. The "NetworkEdgeSecurityService" resource can exist only one per region. It might be the case there is a dangling resource from previous failures. Could you please check if there is such resource in the gcp build project and delete it? The idea is to have a "fresh" env free from dirty resources. I created I step by step here which may help you check that:
1- Query the logs protoPayload.methodName="beta.compute.networkEdgeSecurityServices.insert" AND protoPayload.resourceLocation.currentLocations="asia-southeast1" to identify the possibility of dangling resource and take the resource name.

replace the "BUILD_PROJECT_ID" in the following url:
https://console.cloud.google.com/logs/query;query=protoPayload.methodName%3D%22beta.compute.networkEdgeSecurityServices.insert%22%20AND%0AprotoPayload.resourceLocation.currentLocations%3D%22asia-southeast1%22;cursorTimestamp=2023-08-02T12:22:16.753533Z;duration=P30D?referrer=search&project=BUILD_PROJECT_ID

2- Expand the log and also identify the "request.securityPolicy" which is attached to this service

3- You may use the following APIs to delete both resources:
https://cloud.google.com/compute/docs/reference/rest/v1/networkEdgeSecurityServices/delete

**Comment by zli82016** [comment]
Hello, @felipegc , thanks for the detailed instruction. I deleted both networkEdgeSecurityService and securityPolicy and rerun the tests.

**Comment by zli82016** [comment]
The "NetworkEdgeSecurityService" resource can exist only one per region. 

Does that mean the test can only run once? There is setup to run the test every night. I wonder if the test will fail during the nightly run?

File: mmv1/products/compute/RegionSecurityPolicyRule.yaml [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    # there is a robust handwritten test which covers this scenario. 
+    skip_test: true
+    primary_resource_id: 'policy_rule_network_match'
+    min_version: 'beta'
+    vars:
+      sec_policy_name: 'policyfornetworkmatch'
+parameters:
+  - !ruby/object:Api::Type::String
+    name: 'region'
+    description: |
```
Comment (zli82016) [inline mmv1/products/compute/RegionSecurityPolicyRule.yaml:77]: Should add `immutable: true` to the `region` field? The use case is that when the region changes, should it create a new resource?

File: mmv1/products/compute/RegionSecurityPolicyRule.yaml [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+      If it evaluates to true, the corresponding 'action' is enforced.
+    properties:
+      - !ruby/object:Api::Type::Enum
+        name: 'versionedExpr'
+        description: |
+          Preconfigured versioned expression. If this field is specified, config must also be specified.
+          Available preconfigured expressions along with their requirements are: SRC_IPS_V1 - must specify the corresponding srcIpRange field in config.
+        values:
+          - :SRC_IPS_V1
+        default_value: :SRC_IPS_V1
```
Comment (zli82016) [inline mmv1/products/compute/RegionSecurityPolicyRule.yaml:114]: Is the `default_value` needed here?

File: mmv1/products/compute/RegionSecurityPolicyRule.yaml [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      - !ruby/object:Api::Type::NestedObject
+        name: 'config'
+        description: |
+          The configuration options available when specifying versionedExpr.
+          This field must be specified if versionedExpr is specified and cannot be specified if versionedExpr is not specified.
+        required: true
```
Comment (zli82016) [inline mmv1/products/compute/RegionSecurityPolicyRule.yaml:120]: It looks like the field `config` is only required when `versionedExpr` is specified. `required: true` is not always true.

File: mmv1/products/compute/RegionSecurityPolicyRule.yaml [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+      * allow: allow access to target.
+
+      * deny(STATUS): deny access to target, returns the HTTP response code specified. Valid values for STATUS are 403, 404, and 502.
+
+      * rate_based_ban: limit client traffic to the configured threshold and ban the client if the traffic exceeds the threshold. Configure parameters for this action in RateLimitOptions. Requires rateLimitOptions to be set.
+
+      * redirect: redirect to a different target. This can either be an internal reCAPTCHA redirect, or an external URL-based redirect via a 302 response. Parameters for this action can be configured via redirectOptions. This action is only supported in Global Security Policies of type CLOUD_ARMOR.
+
+      * throttle: limit client traffic to the configured threshold. Configure parameters for this action in rateLimitOptions. Requires rateLimitOptions to be set for this.
+    required: true
```
Comment (zli82016) [inline mmv1/products/compute/RegionSecurityPolicyRule.yaml:141]: Is `action` required?

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+)
+
+func TestAccComputeRegionSecurityPolicyRule_regionSecurityPolicyRuleBasicUpdate(t *testing.T) {
+	t.Parallel()
+
+	context := map[string]interface{}{
+		"random_suffix": RandString(t, 10),
+	}
+
+	VcrTest(t, resource.TestCase{
```
Comment (zli82016) [inline mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:19]: `VcrTest` is deprecated and can be replaced with `acctest.VcrTest`

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: ProtoV5ProviderFactories(t),
```
Comment (zli82016) [inline mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:21]: `ProtoV5ProviderFactories(t)` is deprecated and can be replaced with `acctest.ProtoV5ProviderFactories(t)`

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+}
+`, context)
+}
+
+func TestAccComputeRegionSecurityPolicyRule_regionSecurityPolicyRuleNetworkMatchUpdate(t *testing.T) {
+	context := map[string]interface{}{
+		"random_suffix": RandString(t, 10),
+	}
+
+	VcrTest(t, resource.TestCase{
```
Comment (zli82016) [inline mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:101]: `VcrTest` is deprecated and can be replaced with `acctest.VcrTest`

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: ProtoV5ProviderFactories(t),
```
Comment (zli82016) [inline mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb:103]: `ProtoV5ProviderFactories(t)` is deprecated and can be replaced with `acctest.ProtoV5ProviderFactories(t)`

**Comment by zli82016** [comment]
Thanks for making the changes. Most of this PR looks good. Left a couple of minor comments.

File: mmv1/products/compute/RegionSecurityPolicyRule.yaml (hunk 5 above)
Comment (felipegc) [reply mmv1/products/compute/RegionSecurityPolicyRule.yaml:77]: Done

File: mmv1/products/compute/RegionSecurityPolicyRule.yaml (hunk 8 above)
Comment (felipegc) [reply mmv1/products/compute/RegionSecurityPolicyRule.yaml:140]: It is. The Api returns an error 400 with the following msg: 
"message": "Invalid value for field 'resource.action': ''. Invalid action: "

File: mmv1/products/compute/RegionSecurityPolicyRule.yaml (hunk 6 above)
Comment (felipegc) [reply mmv1/products/compute/RegionSecurityPolicyRule.yaml]: Done

File: mmv1/products/compute/RegionSecurityPolicyRule.yaml (hunk 7 above)
Comment (felipegc) [reply mmv1/products/compute/RegionSecurityPolicyRule.yaml]: Done

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 9 above)
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb]: Done

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 10 above)
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb]: Done

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 11 above)
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb]: Done

File: mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb (hunk 12 above)
Comment (felipegc) [reply mmv1/third_party/terraform/tests/resource_compute_region_security_policy_rule_test.go.erb]: Done

**Comment by felipegc** [comment]
Hi @zli82016 the code is ready to review again. I put this PR in draft mode to avoid merge for now since the user_defined_fields will be public available in few days. The idea is to let this PR ready for merge once it happens so then I move to "wip" again.
Thanks

**Comment by felipegc** [comment]
@zli82016 These features has been moved to "public". Could you please continue the review?

**Review by zli82016 — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**